stackdriver: Refactor special fields docs (#1585)
fluent/fluent-bit-docs: pipeline/outputs/stackdriver_special_fields.md

## Log Entry Fields

Currently, we support some special fields in fluent-bit for setting fields on the LogEntry object:
| JSON log field | [LogEntry](https://cloud.google.com/logging/docs/reference/v2/rest/v2/LogEntry) field | type | Description |
| :--- | :--- | :--- | :--- |

> | `logging.googleapis.com/logName` | `logName` | `string` | The log name to write this log to. |

Reviewer: Should we mention here or in this doc that each of those fields must no exceed a cardinality of 30k (https://cloud.google.com/logging/docs/logs-based-metrics/troubleshooting#too-many-time-series) if one wants to use log based metrics?

Author: The particular way that log name happens to interact with log based metrics strikes me as something Fluent Bit itself shouldn't be concerned about, so I wasn't sure the right way to bring that in. I think I found a nice way to bring it in, while still mostly leaning on linking out to our own docs on the subjects. In [40631d3](https://github.com/fluent/fluent-bit-docs/pull/1585/commits/40631d3fc506102f79ef4b5a4364c5a47d7b75a9) I add a new section to the main stackdriver plugin docs about default Log Name behaviour, along with how to customize it and some examples for why you may want to do this. Hopefully this should get across the point for anyone else who may encounter the same problem.

Side note: I think our own docs on this aren't amazing on this front either, and could probably do with some polishing. I'll raise it with the technical writers who (I think might) work on this internally.

Reviewer: Thanks a lot! An extra section makes a lot of sense.

    "jsonPayload": {
        ...
    }
    "operation": {
        "first": true,
        "last": true
    }
    ...
}
```

> **If the field itself is not a map, the plugin will leave this field untouched.** For example:

Reviewer 2: ```suggestion
If the field itself is not a map, the plugin will leave this field untouched. For example:
```

Author: Fixed in [40631d3](https://github.com/fluent/fluent-bit-docs/pull/1585/commits/40631d3fc506102f79ef4b5a4364c5a47d7b75a9)

```text
{
    "logging.googleapis.com/operation": "some string",
    ...
}
```
the logEntry will be:
```text
{
    "jsonPayload": {
        "logging.googleapis.com/operation": "some string",
        ...
    }
    ...
}
```

> **If there are extra subfields, the plugin will add the recognized fields to the corresponding field in the LogEntry, and preserve the extra subfields in jsonPayload.** For example:

Reviewer 2: ```suggestion
If there are extra subfields, the plugin will add the recognized fields to the corresponding field in the LogEntry, and preserve the extra subfields in jsonPayload. For example:
```

Author: Fixed in [40631d3](https://github.com/fluent/fluent-bit-docs/pull/1585/commits/40631d3fc506102f79ef4b5a4364c5a47d7b75a9)

```text
{
    "jsonPayload": {
        "logging.googleapis.com/operation": {
            "extra1": "some string",
            "extra2": 123,
            "extra3": true
        }
        ...
    }
    "operation": {
        "id": "test_id",
        "producer": "test_producer",
        "first": true,
        "last": true
    }
    ...
}
```

### Exceptions

#### Monitored Resource ID

The `logging.googleapis.com/monitored_resource` field is parsed in a special way, meaning it has some important exceptions:

The `type` field from the [`MonitoredResource` proto]() is not parsed out of the special field. It is read from the [`resource` plugin configuration option](https://docs.fluentbit.io/manual/pipeline/outputs/stackdriver#configuration-parameters). If it is supplied in the `logging.googleapis.com/monitored_resource` special field, it will not be recognized.

> The `labels` field is expected to be an `object<string, string>`. If any fields have a value that is not a string, the value will be ignored and **not preserved**. The plugin will log an error and drop the field.

Reviewer 2: ```suggestion
The `labels` field is expected to be an `object<string, string>`. If any fields have a value that is not a string, the value will be ignored and not preserved. The plugin logs an error and drops the field.
```

Author: Fixed in [f2f67cf](https://github.com/fluent/fluent-bit-docs/pull/1585/commits/f2f67cf3cd9d5af702c78466174accaa9e401754)
